text/0000-build-std/4-stage-1a.md

```diff
@@ -0,0 +1,1001 @@
+# Proposal
+[proposal]: #proposal
+
+Cargo configuration will contain a new key `build-std` under the `[build]`
+section ([?][rationale-build-std-in-config]), permitting one of two values -
+"off" ([?][rationale-build-std-off]) or "always", defaulting to "off":
+
+```toml
+[build]
+build-std = "always" # or `off`
+```
+
+`build-std` can also be specified in the `[target.<triple>]` and
+`[target.<cfg>]` sections ([?][rationale-build-std-target-section]):
+
+```toml
+[target.aarch64-unknown-illumos]
+build-std = "always" # or `off`
+```
+
+The `build-std` configuration locations have the following precedence
+([?][rationale-build-std-precedence]):
+
+1. `[target.<triple>]`
+2. `[target.<cfg>]`
+3. `[build]`
+
+As the Cargo configuration is local to the current installation of Cargo
+(typically in `~/.config/cargo`), the value of `build-std` is not influenced by
+the dependencies of the current crate.
+
+When `build-std` is set to "always", then the standard library will be
+unconditionally recompiled ([?][rationale-unconditional]) in its release profile
+as part of every clean build ([?][rationale-release-profile]). This is primarily
+useful for users of tier three targets.
+
+> [!NOTE]
+>
+> Configuration of the pre-built standard library is split across bootstrap and
+> the Cargo packages for the standard library. As much of this configuration as
+> possible should be moved to the Cargo profile for these packages so that the
+> artifacts produced by build-std match the pre-built standard library as much
+> as is feasible.
+
+Alongside `build-std`, a `build-std-crate` key will be introduced
```

Comment:
`build-std-crate` only really exists because of the staged proposal, and that feels unfortunate, see [#project-goals/2025h1/build-std > targets and stability guarantees](https://rust-lang.zulipchat.com/#narrow/channel/516120-project-goals.2F2025h1.2Fbuild-std/topic/targets.20and.20stability.20guarantees) for a discussion on how we might be able to avoid it.

Reply:
Hmm. I think this is a good option to keep regardless of future extensions as this is a global configuration.

If I set this to just `core` and `build-std` to `always`, then later on if a crate I use depends on alloc or std I should get an error instead of cargo trying to build the "more advanced crates".

Reply:
Isn't this needed for the transition period when crates are still using `#[no_std]` so we don't have a way to detect which crates should be used?

Reply:
> Isn't this needed for the transition period when crates are still using `#[no_std]` so we don't have a way to detect which crates should be used?

That's the idea, yes. I haven't been them most keen on it because it seemed a bit unfortunate to be stuck with both a `build-std-crates` option and the explicit dependencies, being two ways of doing the same thing. But I don't feel too strongly about that, if we think the option has value and is worth keeping regardless then I'm fine to keep it.